x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java

```diff
@@ -47,13 +51,20 @@ public RelevanceMatchQueryRewriter(
         this.relevanceSettingsService = relevanceSettingsService;
         this.curationsService = curationsService;
         this.queryFieldsResolver = queryFieldsResolver;
+        this.queryConfiguration = null;
```

Comment:
The QueryConfiguration was previously retrieved in one place, to get fields and boosts (not my boosts, the other boosts, that App Search calls weights). Now I need it in another place, to get the scoring script (what App Search calls boosts). I don't want an extra request to Elasticsearch to retrieve the configuration, so I made it a member here.

Reply:
I understand the desire not to leave it. However, if multiple callers are calling `RelevanceMatchQueryBuilder` I think this could potentially lead to a concurrency bug.

Reply:
I don't see how multiple calls to `rewriteQuery()` would lead to a concurrency problem. The `QueryConfiguration` is currently retrieved or unset every time `rewriteQuery()` is called. The only difference being, it's retrieved once, not twice. Can you elaborate?

Reply:
The `RelevanceMatchQueryRewriter` is instantiated once, when we create the components for the `RelevanceSearchPlugin`. Then in `getQueries()` we pass the same instance of the `RelevanceMatchQueryRewriter` to every new `RelevanceMatchQueryBuilder` object that we create. This means that every instance of `RelevanceMatchQueryBuilder will share the same `RelevanceMatchQueryRewriter` including all of its state. This is fine for things like the `RelevanceSettingsService`, but the query configuration is unique to a specific query and since the rewriter is shared across all builders, the members of the rewriter can be overwritten by a different builder. This leads to potentially incorrect output when we rewrite the query. A workaround could be to make the query configuration synchronized, but that would introduce blocking. I think what we want to do here is not have shared state that can be overwritten by another instance.

Reply:
Ah, that explains it. I switched from a class member to a local variable.